libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1564,18 +1564,62 @@ bool ExpressionCompiler::visit(IndexAccess const& _indexAccess)
 				setLValueToStorageItem(_indexAccess);
 			break;
 		case DataLocation::Memory:
+			ArrayUtils(m_context).accessIndex(arrayType);
 			setLValue<MemoryItem>(_indexAccess, *_indexAccess.annotation().type, !arrayType.isByteArray());
 			break;
 		case DataLocation::CallData:
-			//@todo if we implement this, the value in calldata has to be added to the base offset
-			solUnimplementedAssert(!arrayType.baseType()->isDynamicallySized(), "Nested arrays not yet implemented.");
-			if (arrayType.baseType()->isValueType())
-				CompilerUtils(m_context).loadFromMemoryDynamic(
-					*arrayType.baseType(),
-					true,
-					!arrayType.isByteArray(),
-					false
+			if (arrayType.baseType()->isDynamicallyEncoded())
+			{
+				// stack layout: <base_ref> <length> <index>
+				ArrayUtils(m_context).accessIndex(arrayType, true, true);
+				// stack layout: <base_ref> <ptr_to_tail>
+
+				// returns the absolute offset of the accessed element in "base_ref"
+				m_context.appendInlineAssembly(
+					Whiskers(R"({
+					let rel_offset_of_tail := calldataload(ptr_to_tail)
+					if iszero(slt(rel_offset_of_tail, sub(sub(calldatasize(), base_ref), 0x1f))) { revert(0, 0) }
+					base_ref := add(base_ref, rel_offset_of_tail)
+				})").render(),
+					{"base_ref", "ptr_to_tail"}
 				);
+				// stack layout: <absolute_offset_of_tail> <garbage>
```

Comment:
It might be a little less optimal, but I think it is better to pop the garbare here right away.

Reply:
Would it help if I reverse the ``if`` cases below instead? Popping here would require to push something back in below for the inline assembly or would there be another way I'm missing?

Reply:
Ah you are right! Yeah then perhaps it's best to leave it like that.

I think we really need a better way to inject inline assembly that has different variables at entry and exit.

Reply:
I think we need to switch entirely to Yul for all this ;-).

Reply:
Now the ``else`` case with the pop comes first - that might be a bit nicer to read.